testsuite/testcases/src/validator_join_leave_test.rs

```diff
@@ -0,0 +1,257 @@
+            validator_cli_indices.push(validator_cli_index);
+            // faucet can make our root LocalAccount sequence number get out of sync.
+            runtime.block_on(async {
+                ctx.swarm()
+                    .chain_info()
+                    .resync_root_account_seq_num(&rest_client)
+                    .await
+                    .unwrap();
+
+                let local_port = port + 1;
+
+                cli.initialize_validator(
+                    validator_cli_index,
+                    keys.consensus_public_key(),
+                    keys.consensus_proof_of_possession(),
+                    HostAndPort {
+                        host: dns_name("0.0.0.0"),
+                        port: local_port,
+                    },
+                    keys.network_public_key(),
+                )
+                .await
+                .unwrap();
+
+                cli.join_validator_set(validator_cli_index, None)
+                    .await
+                    .unwrap();
+
+                reconfig(
+                    &rest_client,
+                    &transaction_factory,
+                    ctx.swarm().chain_info().root_account(),
+                )
+                .await;
+            });
+
+            assert_eq!(
+                runtime.block_on(async { get_validator_state(&cli, validator_cli_index).await }),
+                ValidatorState::ACTIVE
+            );
+        }
+
+        // Log the test setup
+        info!(
+            "Running validator join and leave test {:?} with {:?} validators.",
+            self.name(),
+            num_validators,
+        );
+
+        // Generate some traffic through the validators.
+        // We do this for half the test time.
+        let emit_txn_duration = ctx.global_duration.checked_div(2).unwrap();
+        info!(
+            "Generating the initial traffic for {:?} seconds.",
+            emit_txn_duration.as_secs()
+        );
+        let _txn_stat = generate_traffic(ctx, &all_validators, emit_txn_duration, 1)?;
+
+        // Wait for all nodes to synchronize and stabilize.
+        info!("Waiting for the validators to be synchronized.");
+        runtime.block_on(async {
+            ctx.swarm()
+                .wait_for_all_nodes_to_catchup(Duration::from_secs(MAX_NODE_LAG_SECS))
+                .await
+        })?;
+
+        // Stop and reset 1/3 validators
+        info!("Make 1/3 validators leave the validator set!");
+        for operator_index in validator_cli_indices.iter().take(num_validators / 3) {
+            runtime.block_on(async {
+                cli.leave_validator_set(*operator_index, None)
+                    .await
+                    .unwrap();
+
+                reconfig(
+                    &rest_client,
+                    &transaction_factory,
+                    ctx.swarm().chain_info().root_account(),
+                )
+                .await
+            });
+        }
+
```

Comment:
In addition to this, it would also be good to do some sort of stress testing on nodes joing/leaving validator set. How about we do above operation in loop for several time and at the end you ensure that the validators are able to catch up with the network.

Reply:
Split the test into 3 durations.
With regards to the stress testing, can we do it in another PR and a separate test?

Reply:
A separate PR sounds good to me.